File: python/lira/ir_ser_txt.py
from lira.ir import *

import re

def serialize_shape(shape: Shape) -> str:
    return f'{shape.lanes_base}{shape.lanes_mult if shape.lanes_mult else ""}'

def deserialize_shape(s: str) -> Shape:
    m = re.fullmatch(r'(\d+)(.*)', s)
    return Shape(int(m.group(1)), m.group(2) or None)

def serialize_statement(stmt: Statement) -> str:
    shape_str = serialize_shape(stmt.shape)
    out_parts = [x for typ, name in zip(stmt.outputs_types, stmt.outputs) for x in (str(typ), name)]
    parts = [shape_str] + out_parts + ['=', stmt.kind, stmt.specifier] + stmt.inputs
    return " ".join(parts)

def deserialize_statement(s: str) -> Statement:
    m = re.match(r'^(\w+)\s+(.*?)\s*=\s*(\w+)\s+(\S+)\s*(.*)$', s)
    shape_str, outputs_str, kind, specifier, inputs_str = m.groups()
    shape = deserialize_shape(shape_str)
    pairs = outputs_str.split()
    outputs_types = [int(pairs[i]) for i in range(0, len(pairs), 2)]
    outputs = [pairs[i+1] for i in range(0, len(pairs), 2)]
    inputs = inputs_str.split() if inputs_str else []
    return Statement(shape, outputs, outputs_types, kind, specifier, inputs)

def serialize_statement_seq(seq: StatementSeq) -> str:
    return ''.join(f'{serialize_statement(s)};\n' for s in seq.stmts)

def deserialize_statement_seq(s: str) -> StatementSeq:
    raw_stmts = [part.strip() for part in s.split(";") if part.strip()]
    return StatementSeq([deserialize_statement(raw) for raw in raw_stmts])

File: ruby/lira/ir_ser_txt.rb
require_relative 'ir'

module LIRA
  module IrSerTxt
    def self.serialize_shape(shape)
      "#{shape.lanes_base}#{shape.lanes_mult || ''}"
    end

    def self.deserialize_shape(str)
      if str =~ /^(\d+)(.*)$/
        Shape.new($1.to_i, $2.empty? ? nil : $2)
      else
        raise "Invalid shape string: #{str}"
      end
    end

    def self.serialize_statement(stmt)
      shape_str = serialize_shape(stmt.shape)
      outputs_parts = stmt.outputs_types.zip(stmt.outputs).flat_map { |typ, name| [typ.to_s, name] }
      parts = [shape_str] + outputs_parts + ["=", stmt.kind, stmt.specifier] + stmt.inputs
      parts.join(" ")
    end

    def self.deserialize_statement(str)
      match = str.match(/^(\w+)\s+(.*?)\s*=\s*(\w+)\s+(\S+)\s*(.*)$/)
      raise "Invalid statement: #{str}" unless match
      shape_str, outputs_str, kind, specifier, inputs_str = match.captures
      shape = deserialize_shape(shape_str)
      output_parts = outputs_str.split
      outputs_types = output_parts.each_slice(2).map { |typ, _| typ.to_i }
      outputs = output_parts.each_slice(2).map { |_, name| name }
      inputs = inputs_str.empty? ? [] : inputs_str.split
      Statement.new(shape: shape, outputs: outputs, outputs_types: outputs_types,
                    kind: kind, specifier: specifier, inputs: inputs)
    end

    def self.serialize_statement_seq(seq)
      seq.stmts.map { |s| serialize_statement(s) + ";" }.join("\n")
    end

    def self.deserialize_statement_seq(str)
      raw_stmts = str.split(";").map(&:strip).reject(&:empty?)
      stmts = raw_stmts.map { |raw| deserialize_statement(raw) }
      StatementSeq.new(stmts)
    end
  end
end
